### PR TITLE
Sugar cane aka reeds.

### PR DESCRIPTION
**Pull request by tristanmorgan**
implementing sugar_cane (also know as reeds in the data files). also some small tweaks to soil usage and logic for cactus placement.

### CODE REVIEW

File: server/block/sugar_cane.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import (
+	"github.com/df-mc/dragonfly/server/block/cube"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/df-mc/dragonfly/server/world/particle"
+	"github.com/go-gl/mathgl/mgl64"
+	"math/rand"
+)
+
+// SugerCane is a plant block that generates naturally near water.
```
Comment (JustTalDevelops) [inline server/block/sugar_cane.go:12]:
```suggestion
// SugarCane is a plant block that generates naturally near water.
```

File: server/block/cactus.go [hunk 2]
```diff
@@ -75,6 +61,34 @@ func (c Cactus) RandomTick(pos cube.Pos, w *world.World, r *rand.Rand) {
 	w.SetBlock(pos, Cactus{Age: c.Age}, nil)
 }
 
+// CanGrowHere implements logic to check if cactus can live/grow here
+func (c Cactus) CanGrowHere(pos cube.Pos, w *world.World) bool {
```
Comment (JustTalDevelops) [inline server/block/cactus.go:65]: Out of scope?

Comment (tristanmorgan) [reply server/block/cactus.go:65]: I wanted to D.R.Y.-up the placement logic from the UseOnBlock, NeighbourUpdateTick and RandomTick functions (for both Cane and Cactus). 
@JustTalDevelops I can remove it if that is preferred.

Comment (JustTalDevelops) [reply server/block/cactus.go:65]: Oh that's fair enough, ignore this.

**Review by JustTalDevelops — APPROVED** [review]
Looks good to me if it's all tested. Make sure to run through your changes and correct some docs though, mainly some missing punctuation and capitalization.

File: server/block/cactus.go [hunk 3]
```diff
@@ -75,6 +61,34 @@ func (c Cactus) RandomTick(pos cube.Pos, w *world.World, r *rand.Rand) {
...
+	_, one := w.Block(pos.Side(cube.FaceDown)).(Cactus)
+	if one && supportsVegetation(c, w.Block(pos.Sub(cube.Pos{0, 2}))) {
+		return true
+	}
+
+	// placed on two cactii
+	_, two := w.Block(pos.Side(cube.FaceDown)).(Cactus)
+	if one && two && supportsVegetation(c, w.Block(pos.Sub(cube.Pos{0, 3}))) {
+		return true
+	}
```
Comment (TwistedAsylumMC) [inline server/block/cactus.go:87]: I do not think this is a particularly great idea, why not just do one.CanGrowHere() and allow it to recursively check all the cactus?

File: server/block/sugar_cane.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+		c.Age = 0
+		if c.CanGrowHere(pos.Side(cube.FaceDown), w) {
+			for y := 1; y < 3; y++ {
+				if _, ok := w.Block(pos.Add(cube.Pos{0, y})).(Air); ok {
+					w.SetBlock(pos.Add(cube.Pos{0, y}), SugarCane{Age: 0}, nil)
+					break
+				} else if _, ok := w.Block(pos.Add(cube.Pos{0, y})).(SugarCane); !ok {
+					break
+				}
+			}
```
Comment (TwistedAsylumMC) [inline server/block/sugar_cane.go:57]: Will this not make it grow way too fast? In vanilla only the top sugar cane ever ages, so this logic doesn't exactly match

File: server/block/sugar_cane.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+		}
+	}
+	w.SetBlock(pos, SugarCane{Age: c.Age}, nil)
```
Comment (TwistedAsylumMC) [inline server/block/sugar_cane.go:60]: You can just use c here instead of making a new struct

File: server/block/sugar_cane.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+	_, one := w.Block(pos.Side(cube.FaceDown)).(SugarCane)
+	if one && supportsVegetation(c, w.Block(pos.Sub(cube.Pos{0, 2}))) {
+		return true
+	}
+
+	// placed on two SugarCane
+	_, two := w.Block(pos.Side(cube.FaceDown)).(SugarCane)
+	if one && two && supportsVegetation(c, w.Block(pos.Sub(cube.Pos{0, 3}))) {
+		return true
+	}
```
Comment (TwistedAsylumMC) [inline server/block/sugar_cane.go:85]: Same thing as before with the recursive CanGrowHEre

File: server/block/sugar_cane.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+	return placed(ctx)
+}
+
+// NeighbourUpdateTick ...
+func (c SugarCane) NeighbourUpdateTick(pos, _ cube.Pos, w *world.World) {
+	if !c.CanGrowHere(pos, w) {
+		w.SetBlock(pos, nil, nil)
+		w.AddParticle(pos.Vec3Centre(), particle.BlockBreak{Block: c})
+	}
+}
```
Comment (TwistedAsylumMC) [inline server/block/sugar_cane.go:41]: You also need to drop the item when it breaks

File: server/block/cactus.go [hunk 8]
```diff
@@ -25,31 +25,17 @@ func (c Cactus) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.
...
-			w.AddParticle(pos.Vec3Centre(), particle.BlockBreak{Block: c})
-			return
-		}
-	}
-	_, ok := w.Block(pos.Side(cube.FaceDown)).(Cactus)
-	if !supportsVegetation(c, w.Block(pos.Side(cube.FaceDown))) && !ok {
+	if !c.CanGrowHere(pos, w) {
 		w.SetBlock(pos, nil, nil)
 		w.AddParticle(pos.Vec3Centre(), particle.BlockBreak{Block: c})
 	}
```
Comment (TwistedAsylumMC) [inline server/block/cactus.go:41]: Looks like it was missing before, but could you make it drop the cactus here as well?

**Review by TwistedAsylumMC — COMMENTED** [review]
Looks good, just a few comments

File: server/block/cactus.go (hunk 3 above)
Comment (tristanmorgan) [reply server/block/cactus.go]: changed to a recursive function. limiting the recursion of the height of the plant. (same idea used for sugar cane and cactus.)

File: server/block/sugar_cane.go (hunk 4 above)
Comment (tristanmorgan) [reply server/block/sugar_cane.go:58]: Taken from the Cactus code. it technically only grows when the base is randomly ticked.

File: server/block/cactus.go (hunk 8 above)
Comment (tristanmorgan) [reply server/block/cactus.go]: added dropItem().

File: server/block/sugar_cane.go (hunk 7 above)
Comment (tristanmorgan) [reply server/block/sugar_cane.go]: drops item now.

File: server/block/sugar_cane.go (hunk 6 above)
Comment (tristanmorgan) [reply server/block/sugar_cane.go]: done.

File: server/block/sugar_cane.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+	Age int
+}
+
+// UseOnBlock ensures the placement of the block is OK.
+func (c SugarCane) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) (used bool) {
+	pos, _, used = firstReplaceable(w, pos, face, c)
+	if !used {
+		return false
+	}
+	if !c.CanGrowHere(pos, w, 3) {
```
Comment (DaPigGuy) [inline server/block/sugar_cane.go:27]: This isn't quite right as you can continue stacking sugarcane through placement. In fact, sugar cane in vanilla Minecraft can naturally generate 4 blocks tall!

File: server/block/sugar_cane.go [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+		return false
+	}
+
+	place(w, pos, c, user, ctx)
+	return placed(ctx)
+}
+
+// NeighbourUpdateTick ...
+func (c SugarCane) NeighbourUpdateTick(pos, _ cube.Pos, w *world.World) {
+	if !c.CanGrowHere(pos, w, 3) {
```
Comment (DaPigGuy) [inline server/block/sugar_cane.go:37]: See above comment on height.

File: server/block/sugar_cane.go [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+				} else if _, ok := w.Block(pos.Add(cube.Pos{0, y})).(SugarCane); !ok {
+					break
+				}
+			}
+		}
+	}
+	w.SetBlock(pos, c, nil)
+}
+
+// CanGrowHere implements logic to check if sugar cane can live/grow here.
```
Comment (DaPigGuy) [inline server/block/sugar_cane.go:64]: Any reason to have this exported?

**Review by DaPigGuy — CHANGES_REQUESTED** [review]

File: server/block/sugar_cane.go (hunk 9 above)
Comment (tristanmorgan) [reply server/block/sugar_cane.go]: I have removed the height checking to allow higher placements.

File: server/block/sugar_cane.go (hunk 11 above)
Comment (tristanmorgan) [reply server/block/sugar_cane.go]: n00b mistake. fixed.

File: server/block/sugar_cane.go (hunk 10 above)
Comment (tristanmorgan) [reply server/block/sugar_cane.go]: changed.

**Review by DaPigGuy — CHANGES_REQUESTED** [review]
PR is currently missing bone meal behavior found in Bedrock Edition

File: server/block/cactus.go [hunk 12]
```diff
@@ -25,33 +25,20 @@ func (c Cactus) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.
 	if !used {
 		return false
 	}
-	_, ok := w.Block(pos.Side(cube.FaceDown)).(Cactus)
-	if !supportsVegetation(c, w.Block(pos.Side(cube.FaceDown))) && !ok {
+	if !c.canGrowHere(pos, w, 2) {
```
Comment (DaPigGuy) [inline server/block/cactus.go:28]: This was missed in the previous review but the total cactus height should not be considered here either.

File: server/block/cactus.go [hunk 13]
```diff
@@ -25,33 +25,20 @@ func (c Cactus) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.
...
-	for _, face := range cube.HorizontalFaces() {
-		if _, ok := w.Block(pos.Side(face)).(Air); !ok {
-			w.SetBlock(pos, nil, nil)
-			w.AddParticle(pos.Vec3Centre(), particle.BlockBreak{Block: c})
-			return
-		}
-	}
-	_, ok := w.Block(pos.Side(cube.FaceDown)).(Cactus)
-	if !supportsVegetation(c, w.Block(pos.Side(cube.FaceDown))) && !ok {
+	if !c.canGrowHere(pos, w, 2) {
```
Comment (DaPigGuy) [inline server/block/cactus.go:38]: Read above comment on cactus height

File: server/block/register.go [hunk 14]
```diff
@@ -177,6 +177,7 @@ func init() {
 	registerAll(allStainedTerracotta())
 	registerAll(allStairs())
 	registerAll(allStoneBricks())
+	registerAll(allSugarCane())
 	registerAll(allStonecutters())
```
Comment (DaPigGuy) [inline server/block/register.go:181]:
```suggestion
	registerAll(allStonecutters())
	registerAll(allSugarCane())
```

Sort alphabetically

File: server/block/sugar_cane.go [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+	"github.com/df-mc/dragonfly/server/block/cube"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/df-mc/dragonfly/server/world/particle"
+	"github.com/go-gl/mathgl/mgl64"
+	"math/rand"
+)
+
+// SugarCane is a plant block that generates naturally near water.
+type SugarCane struct {
```
Comment (DaPigGuy) [inline server/block/sugar_cane.go:13]: SugarCane should implement LiquidRemovable

**Review by DaPigGuy — CHANGES_REQUESTED** [review]

File: server/block/sugar_cane.go (hunk 15 above)
Comment (tristanmorgan) [reply server/block/sugar_cane.go:13]: server/block/crop.go:40:15: has another declaration of HasLiquidDrops.

File: server/block/register.go (hunk 14 above)
Comment (tristanmorgan) [reply server/block/register.go]: done.

File: server/block/cactus.go (hunk 13 above)
Comment (tristanmorgan) [reply server/block/cactus.go]: done.

File: server/block/cactus.go (hunk 12 above)
Comment (tristanmorgan) [reply server/block/cactus.go]: height excluded from check.

File: server/block/sugar_cane.go (hunk 15 above)
Comment (DaPigGuy) [reply server/block/sugar_cane.go:13]: SugarCane is not embedding crop, nor should it

Comment (tristanmorgan) [reply server/block/sugar_cane.go:13]: ahh, found my mistake. added the func now.

**Review by DaPigGuy — APPROVED** [review]

**Review by JustTalDevelops — APPROVED** [review]
Looks good to me!

File: server/block/sugar_cane.go [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+// BoneMeal ...
+func (c SugarCane) BoneMeal(pos cube.Pos, w *world.World) bool {
+	for _, ok := w.Block(pos.Side(cube.FaceDown)).(SugarCane); ok; _, ok = w.Block(pos.Side(cube.FaceDown)).(SugarCane) {
+		pos = pos.Side(cube.FaceDown)
+	}
+	if c.canGrowHere(pos.Side(cube.FaceDown), w, false) {
+		for y := 1; y < 3; y++ {
+			if _, ok := w.Block(pos.Add(cube.Pos{0, y})).(Air); ok {
+				w.SetBlock(pos.Add(cube.Pos{0, y}), SugarCane{Age: 0}, nil)
```
Comment (Sandertv) [inline server/block/sugar_cane.go:72]: Don't need to have `Age: 0` here. That's the default value.

File: server/block/sugar_cane.go [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+	if _, ok := w.Block(pos.Side(cube.FaceDown)).(SugarCane); ok && recursive {
+		return c.canGrowHere(pos.Side(cube.FaceDown), w, recursive)
+	}
+
+	if supportsVegetation(c, w.Block(pos.Sub(cube.Pos{0, 1}))) {
+		for _, face := range cube.HorizontalFaces() {
+			if liquid, ok := w.Liquid(pos.Side(face).Side(cube.FaceDown)); ok {
+				if _, ok := liquid.(Water); ok {
+					return true
+				}
```
Comment (Sandertv) [inline server/block/sugar_cane.go:91]: Can change this to:
```go
_, ok := liquid.(Water)
return ok
```

**Review by Sandertv — APPROVED** [review]
This is pretty much ready to go. I've got 2 small nitpicks but once those are resolved this should be ready to go!

**Pull request merged**